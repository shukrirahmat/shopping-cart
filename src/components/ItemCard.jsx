import PropTypes, { func } from "prop-types";
import styles from "../styles/ItemCard.module.css";
import { useState } from "react";

function ItemCard({ item , addToCart}) {
  const [amount, setAmount] = useState(1);

  function changeAmount(e) {
    const value = e.target.value.replace(/[^0-9]/g, "");
    let num;
    if (!value) num = 0; 
    else num = parseInt(value);
    setAmount(num);
  }

  function handleBlur() {
    if (!amount || amount < 1) setAmount(1);
    if (amount > 99) setAmount(99);
  }

  function increaseAmount() {
    if (amount < 99) setAmount(amount + 1);
  }

  function decreaseAmount() {
    if (amount > 1) setAmount(amount - 1);
  }

  function handleAdd() {
    addToCart(item, amount);
  }

  function pricetoString(price) {
    return price.toFixed(2) + "$";
  }

  return (
    <div className={styles.card}>
      <p className={styles.cardTitle}>{item.title}</p>
      <img src={item.image}></img>
      <div className={styles.priceandamount}>
        <p role="priceTag" className={styles.cardPrice}>{pricetoString(item.price * amount)}</p>
        <div className={styles.amount}>
          <button onClick={decreaseAmount}>-</button>
          <input type="text" value={amount} onBlur={handleBlur} onChange={changeAmount}></input>
          <button onClick={increaseAmount}>+</button>
        </div>
      </div>
      <button className={styles.addtoCart} onClick={handleAdd}>ADD TO CART</button>
    </div>
  );
}

ItemCard.proptypes = {
  item: PropTypes.object,
  addToCart: PropTypes.func,
};

export default ItemCard;
