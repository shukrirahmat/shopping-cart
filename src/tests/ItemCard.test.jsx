import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ItemCard from "../components/ItemCard";

describe("Item card component", () => {
    const item = {
        id: '69',
        title: 'test product',
        price: 13.5,
        description: 'lorem ipsum set',
        image: 'https://i.pravatar.cc',
        category: 'electronic'
    }

    it("renders card", () => {
        const {container} = <ItemCard item={item} key={item.id} addToCart={() => {}}/>
        expect(container).toMatchSnapshot();
    })

    it("increase amount when clicking +", async () => {
        const user = userEvent.setup();

        render(<ItemCard item={item} key={item.id} addToCart={() => {}}/>);
        const input = screen.getByRole("textbox");
        const button = screen.getByRole("button", {name: "+"});
        await user.click(button);
        await user.click(button);
        await user.click(button);

        expect(input.value).toMatch(/4/);
    })

    it("it decreases value when clicking -", async () => {
        const user = userEvent.setup();

        render(<ItemCard item={item} key={item.id} addToCart={() => {}}/>);
        const input = screen.getByRole("textbox");
        const button = screen.getByRole("button", {name: "-"});
        
        await user.clear(input);
        await user.type(input, "10");
        await user.click(button);
        await user.click(button);
        expect(input.value).toMatch(/8/);

    })

    it("updates price when changing amount", async () => {
        const user = userEvent.setup();

        render(<ItemCard item={item} key={item.id} addToCart={() => {}}/>);
        const input = screen.getByRole("textbox");
        const price = screen.getByRole("priceTag");
        
        expect(price.textContent).toMatch("13.50$");

        await user.clear(input);
        await user.type(input, "10")

        expect(price.textContent).toMatch("135.00$");

    })

    it("filters alphabet in input", async () => {
        const user = userEvent.setup();
        render(<ItemCard item={item} key={item.id} addToCart={() => {}}/>);
        const input = screen.getByRole("textbox");
        
        await user.clear(input);
        await user.type(input, "1a2b");
        await user.tab();
        
        expect(input.value).toMatch(/12/);

    })

    it("fixes the amount when value is empty", async () => {
        const user = userEvent.setup();
        render(<ItemCard item={item} key={item.id} addToCart={() => {}}/>);
        const input = screen.getByRole("textbox");
        
        await user.clear(input);
        await user.tab();
        
        expect(input.value).toMatch(/1/);

    })

    it("fixes the amount when value is below 1", async () => {
        const user = userEvent.setup();
        render(<ItemCard item={item} key={item.id} addToCart={() => {}}/>);
        const input = screen.getByRole("textbox");
        
        await user.clear(input);
        await user.type(input, "0");
        await user.tab();
        
        expect(input.value).toMatch(/1/);
    })

    it("fixes the amount when value is higher than 99", async () => {
        const user = userEvent.setup();
        render(<ItemCard item={item} key={item.id} addToCart={() => {}}/>);
        const input = screen.getByRole("textbox");
        
        await user.clear(input);
        await user.type(input, "1000");
        await user.tab();
        
        expect(input.value).toMatch(/99/);
    })

    it("should call the addToCart function when clicked", async () => {
        const addToCart = vi.fn();
        const user = userEvent.setup();
        render(<ItemCard item={item} key={item.id} addToCart={addToCart}/>);
        const button = screen.getByRole("button", {name: /ADD TO CART/});

        await user.click(button);

        expect(addToCart).toHaveBeenCalled();

    })

    it("should not call the addToCart function when not clicked", async () => {
        const addToCart = vi.fn();
        render(<ItemCard item={item} key={item.id} addToCart={addToCart}/>);

        expect(addToCart).not.toHaveBeenCalled();

    })
})