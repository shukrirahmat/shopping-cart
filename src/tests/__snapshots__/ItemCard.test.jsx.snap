// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item card component > renders card 1`] = `undefined`;
